File: src/ca/qc/bdeb/inf203/model/PowerUp.java
package ca.qc.bdeb.inf203.model;

import java.awt.Point;
import java.awt.Rectangle;

/**
 * Power-up que le joueur peut rammasser dans la grille de jeu.
 *
 * @author Nicolas Hurtubise, Guillaume Riou
 */
public abstract class PowerUp extends Entite implements Cloneable {

    protected int animationFrameRate;
    protected long derniereAnimationTimestamp;
    protected int nbrImagesAnimation;
    protected long dernierTicTimestamp = System.currentTimeMillis();
    protected int vitesse;
    protected float pendingDeplacementX;
    protected float pendingDeplacementY;
    protected Point destination = null;

    public PowerUp(String path[]) {
        this.etat = Etat.ATTENTE;
        this.sprites.put(Etat.ATTENTE, new RepresentationImage(path));
    }

    public PowerUp(String path) {
        this.etat = Etat.ATTENTE;
        this.hitbox = new Rectangle();
        this.sprites.put(Etat.ATTENTE, new RepresentationImage(new String[]{"powerups", path}));
    }

    @Override
    public Etat getEtat() {
        return Etat.ATTENTE;
    }

    public int getVitesse() {
        return vitesse;
    }

    public Point getDestination() {
        return destination;
    }

    public void tic() {
        long temps = System.currentTimeMillis();

        if (destination == null || hitbox.getLocation().equals(destination)) {
            // Si le power-up est déjà arrivé à sa destination, rien à faire
            return;
        }

        long deltaTemps = temps - this.dernierTicTimestamp;

        // v = dx/dT => dx = v*dt
        float deltaX = (Math.abs(vitesse) * Math.signum(destination.x - hitbox.getLocation().x) * (deltaTemps / 1000.0f));
        float deltaY = (Math.abs(vitesse) * Math.signum(destination.y - hitbox.getLocation().y) * (deltaTemps / 1000.0f));

        pendingDeplacementX += deltaX;
        pendingDeplacementY += deltaY;

        int deplacementX = (int) pendingDeplacementX;
        int deplacementY = (int) pendingDeplacementY;

        this.hitbox.x += deplacementX;
        this.hitbox.y += deplacementY;

        pendingDeplacementX -= deplacementX;
        pendingDeplacementY -= deplacementY;

        this.dernierTicTimestamp = temps;
    }

    public abstract void action();

    @Override
    public int getAnimationCompteur() {
        long ts = System.currentTimeMillis();

        if (ts - this.derniereAnimationTimestamp >= (1000.0 / (float) animationFrameRate)) {
            // On incrémente l'animation
            animationCompteur++;
            derniereAnimationTimestamp = ts;
        }

        return animationCompteur % nbrImagesAnimation;
    }

    @Override
    protected PowerUp clone() throws CloneNotSupportedException {
        PowerUp clone = (PowerUp) super.clone();
        clone.destination = (Point) this.destination.clone();

        return clone;
    }
}

File: assets/js/model/power-up.js
/**
 * Power-up que le joueur peut rammasser dans la grille de jeu.
 *
 * @author Nicolas Hurtubise, Guillaume Riou
 */
function PowerUp() {
    Entite.apply(this);
    /**
     * Vitesse de l'animation
     */
    this.animationFrameRate = null;
    /**
     * dernier temps calculé pour l'animation.
     */
    this.derniereAnimationTimestamp = null;
    /**
     * Nombre d'images par animation (franchement, faut vraiment faire de la
     * doc pour des trucs comme ça  ? ...)
     */
    this.nbrImagesAnimation = null;
    /**
     * Dernier temps calculé pour une action.
     */
    this.dernierTicTimestamp = window.performance.now();
    /**
     * Vitesse de déplacement.
     */
    this.vitesse = 0;
    /**
     * déplacements fractionnaire en attente.
     */
    this.pendingDeplacementX = 0;
    this.pendingDeplacementY = 0;
    /**
     * Point où le powerUp arrête de tomber.
     */
    this.destination = null;

    this.etat = navigator.Etat.ATTENTE;
    this.hitbox = new Rectangle();
}

PowerUp.prototype = new Entite();

PowerUp.prototype.getEtat = function() {
    return navigator.Etat.ATTENTE;
};

/**
 * Gère les actions des powerups.
 */
PowerUp.prototype.tic = function() {
    var temps = window.performance.now();

    if (this.destination == null || (this.hitbox.x == this.destination.x && this.hitbox.y == this.destination.y)) {
        // Si le power-up est déjà arrivé à sa destination, rien à faire
        return;
    }

    var deltaTemps = temps - this.dernierTicTimestamp;

    // v = dx/dT => dx = v*dt
    var deltaX = (Math.abs(vitesse) * Math.sign(this.destination.x - this.hitbox.x) * (deltaTemps / 1000.0));
    var deltaY = (Math.abs(vitesse) * Math.sign(this.destination.y - this.hitbox.y) * (deltaTemps / 1000.0));

    this.pendingDeplacementX += deltaX;
    this.pendingDeplacementY += deltaY;

    var deplacementX = parseInt(pendingDeplacementX);
    var deplacementY = parseInt(pendingDeplacementY);

    this.hitbox.x += deplacementX;
    this.hitbox.y += deplacementY;

    pendingDeplacementX -= deplacementX;
    pendingDeplacementY -= deplacementY;

    this.dernierTicTimestamp = temps;
};

PowerUp.prototype.action = function() {};

PowerUp.prototype.getAnimationCompteur = function() {
    var ts = window.performance.now();

    if (ts - this.derniereAnimationTimestamp >= (1000 / this.animationFrameRate)) {
        // On incrémente l'animation
        this.animationCompteur++;
        this.derniereAnimationTimestamp = ts;
    }

    return this.animationCompteur % this.nbrImagesAnimation;
};

PowerUp.prototype.clone = function() {
    return clone_json(this);
};
